spring-cloud-data-core/src/main/java/org/springframework/cloud/data/core/ModuleKey.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2015 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.data.core;
+
+import java.io.Serializable;
+
+import org.springframework.util.Assert;
+
+/**
+ * Unique identifier for a {@link ModuleDeploymentRequest}.
+ * Methods {@link #toUID()} and {@link #fromUID(String)} can be used to convert
+ * a key to a string and a string to a key, respectively. The UID string may be
+ * used to uniquely identify a module in a database or execution environment.
+ *
+ * @author Patrick Peralta
+ */
+public class ModuleKey implements Serializable {
+	private static final long serialVersionUID = 1L;
+
+	/**
+	 * The name of the associated group or deployment unit this
+	 * module belongs to, such as a stream or job.
+	 */
+	private final String group;
+
+	/**
+	 * The label provided to uniquely identify the module as part
+	 * of group membership. For instance, if a group has multiple
+	 * {@code filter} modules, each will have a unique label.
+	 */
+	private final String label;
+
+	/**
+	 * Construct a {@code ModuleKey}.
+	 *
+	 * @param group name of group (stream/job) this module belongs to
+	 * @param label label to uniquely identify this module in its group
+	 */
+	public ModuleKey(String group, String label) {
```

Comment:
Yeah I think it will always either be from `ModuleDefinition` or parsing of an existing key. We can make the constructor private.